### PR TITLE
Introduces AlertScreen, MainAndModalContainer.

### LINKED ISSUE #37: AlertDialog integration or replacement
AC

 * Workflow apps can drive AlertDialog or, more likely, something that looks like AlertDialog

### PR DESCRIPTION
**Pull request by rjrjr**
Also:

 * Kills off the bespoke ShellCoordinator.
 * Renames some of the existing classes and methods.
 * Moves backstack stuff into its own package.
 * Moves to appcompatv7 dialogs.

Still no support for custom modals, but it's now pretty obvious where to fill in that blank.

Closes #37.

### CODE REVIEW

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+          for ((i, modalScreen) in mainAndModal.modals.withIndex()) {
+            newDialogs +=
+              if (dialogs.size < i && dialogs[i].screen::class == modalScreen::class) {
+                dialogs[i]
+              } else {
+                DialogRef(modalScreen, mainAndModal.showDialog(i, screens, viewRegistry, this))
+              }
+          }
+
+          dialogs.filter { it !in newDialogs }
```
Comment (zach-klippenstein) [inline kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Nit: Could also do `(dialogs - newDialogs)` I believe, slightly more concise and I think just as expressive. Also, the kotlin runtime actually tries to optimize it a little if you use minus (TIL myself).

Here's the optimizations, if you're curious:
image
image

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        NEGATIVE -> builder.setNegativeButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEGATIVE))
+        }
+        NEUTRAL -> builder.setNeutralButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEUTRAL))
+        }
+      }
+    }
+
+    alertScreen.message.takeIf { it.isNotEmpty() }
```
Comment (zach-klippenstein) [inline kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: `isNotBlank()`?

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        .let { builder.setMessage(it) }
+    alertScreen.title.takeIf { it.isNotEmpty() }
+        .let { builder.setTitle(it) }
+
+    return builder.show()
+  }
+
+  private fun <T> Observable<T>.whileAttached(): Observable<T> =
+    attached.switchMap { isAttached -> if (isAttached) this else never<T>() }
```
Comment (zach-klippenstein) [inline kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Cute

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    mainAndModalScreens: Observable<out MainAndModalScreen<*, *>>,
+    viewRegistry: ViewRegistry,
+    container: ViewGroup
+  ): Dialog {
+    (modals[index] as? AlertScreen)?.let { return showAlert(it) }
+
+    // https://github.com/square/workflow/issues/58
+    TODO("Non-alerts are not supported yet.")
+//    val modalScreens: Observable<out D> = mainAndModalScreens.mapToModalIndexMatching(index, this)
+//    val binding:
```
Comment (zach-klippenstein) [inline kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Nit: I assume you meant to leave this code in here? Want to put it in #58 instead?

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  companion object : ViewBinding<MainAndModalScreen<*, *>>
+  by BuilderBinding(
+      type = MainAndModalScreen::class.java,
+      builder = { screens, builders, context, _ ->
+        MainAndModalContainer(context).apply {
+          layoutParams = (ViewGroup.LayoutParams(
+              android.view.ViewGroup.LayoutParams.MATCH_PARENT,
```
Comment (zach-klippenstein) [inline kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Nit: Can we static-import these?

File: kotlin/sample-game-common/src/main/java/com/squareup/sample/tictactoe/RunGameRenderer.kt [hunk 6]
```diff
@@ -45,7 +54,23 @@ object RunGameRenderer :
...
-          ConfirmQuitScreen(workflow::sendEvent)
+          AlertScreen(
+              workflow.adaptEvents<AlertScreen.Event, RunGameEvent> { alertEvent ->
+                when (alertEvent) {
+                  is ButtonClicked -> when (alertEvent.button) {
+                    POSITIVE -> ConfirmQuit
+                    NEGATIVE -> ContinuePlaying
+                    NEUTRAL -> throw IllegalArgumentException()
+                  }
+                  Canceled -> ContinuePlaying
```
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply kotlin/sample-game-common/src/main/java/com/squareup/sample/tictactoe/RunGameRenderer.kt]: Ah I see it's bound to cancelation instead. That makes much more sense. POS code has conditioned me to expect dismissal.

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt (hunk 1 above)
Comment (zach-klippenstein) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Internal-only slack link unfortunately, but [some interesting discussion of this technique ensued](https://square.slack.com/archives/C0RN3QDN0/p1550099293015000).

Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: I had no idea that was even an option!

```
          (dialogs - newDialogs).forEach { it.dialog.hide() }
          dialogs = newDialogs
```

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt (hunk 3 above)
Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: `switchMap()` seems like the solution to all the problems.

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt (hunk 4 above)
Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: Oops, meant to delete that.

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt (hunk 5 above)
Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: I dropped the `android.view`, but left in `ViewGroup.LayoutParams`. IIRC you need to be _very_ careful to get the right type w/these things.

Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt]: NM, that's covered by the constructor. Imported `MATCH_PARENT`.

File: kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/backstack/BackStackContainer.kt [hunk 7]
```diff
@@ -22,14 +22,22 @@ import android.view.View
...
+import com.squareup.viewregistry.BuilderBinding
+import com.squareup.viewregistry.HandlesBack
+import com.squareup.viewregistry.ViewBinding
+import com.squareup.viewregistry.ViewRegistry
+import com.squareup.viewregistry.backstack.ViewStateStack.SavedState
+import com.squareup.viewregistry.takeWhileAttached
 import io.reactivex.Observable
 
 /**
  * A container view that can display a stream of [BackStackScreen] instances.
```
[the remark this answers is not in the document]
Comment (rjrjr) [reply kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/backstack/BackStackContainer.kt]: He said yes. Which is nice.

**Pull request merged**